vm/state_migration/src/nv12/miner.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! This module implements network version 12 or actorv4 state migration
+//! Please read https://filecoin.io/blog/posts/filecoin-network-v12/
+//! to learn more about network version 12 migration.
+//! This is more or less a direct port of the state migration
+//! implemented in lotus' specs-actors library.
+
+use crate::{ActorMigration, ActorMigrationInput};
+use crate::{MigrationError, MigrationOutput, MigrationResult};
+use actor_interface::actorv3::miner::State as V3State;
+use actor_interface::actorv4::miner::State as V4State;
+use async_std::sync::Arc;
+use cid::Cid;
+use cid::Code::Blake2b256;
+use ipld_blockstore::BlockStore;
+
+pub struct MinerMigrator(Cid);
+
+pub fn miner_migrator_v4<BS: BlockStore + Send + Sync>(
+    cid: Cid,
+) -> Arc<dyn ActorMigration<BS> + Send + Sync> {
+    Arc::new(MinerMigrator(cid))
+}
+
+// each actor's state migration is read from blockstore, changes state tree, and writes back to the blocstore.
+impl<BS: BlockStore + Send + Sync> ActorMigration<BS> for MinerMigrator {
+    fn migrate_state(
+        &self,
+        store: Arc<BS>,
+        input: ActorMigrationInput,
+    ) -> MigrationResult<MigrationOutput> {
+        let v3_state: Option<V3State> = store
+            .get(&input.head)
+            .map_err(|e| MigrationError::BlockStoreRead(e.to_string()))?;
```

Comment:
I think you can chain the `ok_or_else` to the `map_err` if you add a `?` after `map_err`